internal/shaderir/check.go

```diff
@@ -98,6 +111,13 @@ func TypeFromBinaryOp(op Op, lhst, rhst Type, lhsConst, rhsConst constant.Value)
 			return Type{}, false
 		}
 
+		if op == LeftShift || op == RightShift {
+			if lhsConst.Kind() == constant.Int && rhsConst.Kind() == constant.Int {
+				return Type{Main: Int}, true
+			}
+			return Type{}, false
```

Comment:
Even if one of the consts are typed, if both can be truncated to integers, this should be fine (https://go.dev/play/p/o5hJUTxFbYK). Does this test pass?

```go
{stmt: "1.0 << int(1)", err: false}
{stmt: "int(1) << 1.0", err: false}
```

Reply:
Yes, if one of the constants is typed, the second will be deduced similarly to other operators. The only difference is when both of them are untyped const, operands will be treated as integers.
```go
{stmt: "a := 1.0 << int(2); _ = a", err: false}, // 1.0 becomes int
{stmt: "a := int(1) << 2.0; _ = a", err: false}, // 2.0 becomes int
{stmt: "a := int(1) << 2.0; _ = a", err: }, // 2.0 become sint
{stmt: "a := 1.0 << 2.0; _ = a", err: false}, // Both are ints
{stmt: "a := 1.0 << float(2.0); _ = a", err: true}, // 1.0 becomes float, TypeFromBinaryOp will fail for float << float
```